### PR TITLE
Make Ray container image build reproducible

### PR DESCRIPTION
**Pull request by astefanutti**
## Description

This PR updates the Ray container images build to use `Pipefile.lock` files so it's reproducible. 

See https://issues.redhat.com/browse/RHOAIENG-11120.

## How Has This Been Tested?

I've successfully tested the [LLMs fine-tuning with Ray example](https://github.com/opendatahub-io/distributed-workloads/tree/main/examples/ray-finetune-llm-deepspeed) using images built with these changes.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: images/runtime/examples/ray-torch/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,5 @@
 FROM quay.io/rhoai/ray:2.23.0-py39-cu121
 
-RUN python3 -m pip install --no-cache-dir -U pip \
-    torch==2.3.1
+COPY Pipfile Pipfile.lock ./
+
+RUN pipenv install --system
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply images/runtime/examples/ray-torch/Dockerfile:5]: Once that PR is merged and the new base container image pushed, it'll come with that base image so custom images built on top of that base image will have pipenv installed already.

**Review by sutaakar — COMMENTED** [review]
/lgtm

**Review by Ygnas — APPROVED** [review]

**Review by Ygnas — APPROVED** [review]

**Review by KPostOffice — COMMENTED** [review]
There's some tricks here with multistage build so that we don't need to include pipenv and setuptools in the final image

https://pipenv.pypa.io/en/stable/docker.html

**Comment by astefanutti** [comment]
> There's some tricks here with multistage build so that we don't need to include pipenv and setuptools in the final image
> 
> https://pipenv.pypa.io/en/stable/docker.html

@KPostOffice I've been trying the multistage approach, but it feels like it adds too much complexity for the benefits. I've updated this PR to use https://github.com/thoth-station/micropipenv, which is what the notebooks team relies on, and somehow addresses some of the concerns.

**Review by KPostOffice — APPROVED** [review]
/lgtm

**Pull request merged**